{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 1"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hidden Messages in the Replication Origin"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How does the bacterial cell know to begin replication exactly in this short region within the much larger Vibrio cholerae chromosome, which consists of 1,108,250 nucleotides?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PatternCount(text, pattern):\n",
    "\n",
    "    count = 0\n",
    "\n",
    "    for i in range(len(text) - len(pattern) + 1):\n",
    "        if text[i:i+len(pattern)] == pattern:\n",
    "            count += 1\n",
    "            \n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome = 'CCCTCAATAGGCTCAATACTCAATACTCAATACCTCAATAGGCCTCAATACTCCTCAATATCTCAATAGCCTCAATACCTCAATACTCAATACTCAATACCCCTCAATATAATAGCTCAATACATCTCAATAGTGTCTCAATAGCGCGACTCTCAATATCTCAATATACTCAATACTCAATATGCAGTCTCAATACTCAATAACTCAATACTCAATAACTCAATAAGCTCAATACGCACTCAATACTCAATAGACTCTCAATACTCAATACCCTCAATACTCAATACTCAATACCTCAATAAAGTTCTCAATACTCAATACTCAATACTCAATAACCCTCAATAGCTCAATAGGCTCAATACTCGTCTCAATATCTCAATAAGTCTCAATACTCAATACTCAATACAGCTCAATAACCTCAATATTCTCAATACTCAATAACTCAATAATCTTCTCAATACTCAATAAATATACCGGCTCAATAGGTGCTCAATACTCAATAACTAGCTCAATACGGTTACTCAATAGATTCTCAATAACTCAATAGATGCTCAATAATCCTCAATACTCAATACTCAATAATGGCTCAATACGGGCTCAATACTCAATACTCAATAACTCAATACTCAATATTCAGCTCAATATCGTTCTCTCAATACACCTCAATACTCAATAGACACTCAATACTCAATAGGCATCTCAATATCATCTCAATAACTCAATACTCTCAATAATCCTCAATATCTCAATATCCTCAATAGGCGGCTCAATAATTTCCCTCTTGAACCTCAATACCTCAATACCTCAATATACGGGCCTCAATAGCGGTATACGGGTACTCAATACGACTCCTCAATAACCTCAATAGCTCAATAAACTCAATATGGCTCAATATCCTCAATAATACAGCGAGCTCAATAACTCAATACTCAATAATTTGACCTCAATACTCAATACTCAATACTCAATAAACTCAATACAATCCTCAATA'\n",
    "pattern = 'CTCAATACT'\n",
    "PatternCount(genome,pattern)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A Faster Frequent Words Approach"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def FrequencyTable(Text, k):\n",
    "    '''\n",
    "    Function to create dictionary of all k-mers and their counts in chosen Text\n",
    "    '''\n",
    "    freqMap = dict()\n",
    "\n",
    "    for i in range(len(Text) - k + 1):\n",
    "        Pattern = Text[i:i+k]\n",
    "        if Pattern not in freqMap:\n",
    "            freqMap[Pattern] = 1\n",
    "        else:\n",
    "            freqMap[Pattern] += 1\n",
    "    return freqMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def SortPatternsDict(Dictionary):\n",
    "    '''\n",
    "    Function return dictionary sorted by value.\n",
    "    '''\n",
    "\n",
    "    sortedDict = {\n",
    "        key: value for key, value in sorted(Dictionary.items(), key=lambda item: item[1], reverse=True)\n",
    "    }\n",
    "    return sortedDict\n",
    "    \n",
    "    \n",
    "def MaxFreq(Dictionary: dict):\n",
    "    '''\n",
    "    Function to return max value from dictionary\n",
    "    '''\n",
    "\n",
    "    maxValue = max(\n",
    "        Dictionary, key=Dictionary.get\n",
    "    )\n",
    "    return Dictionary[maxValue]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def BetterFrequentWords(Text, k):\n",
    "    '''\n",
    "    Function to find the most common k-mers of certain length in text\n",
    "    '''\n",
    "    \n",
    "    frequentPatterns = list()\n",
    "    freqMap = FrequencyTable(Text, k)\n",
    "\n",
    "    max_freq = MaxFreq(freqMap)\n",
    "\n",
    "    for pattern in freqMap:\n",
    "        if freqMap[pattern] == max_freq:\n",
    "            frequentPatterns.append(pattern)\n",
    "\n",
    "    return frequentPatterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AAACGTGCAAA']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BetterFrequentWords(\n",
    "    'CCTCCAAGGCGGGTACGTGCAAAACGTGCAAAGGCGGGTATCGAACCGGCGGGTATCGAACCACGTGCAAACCTCCAATGACGCAATGACGCAAGGCGGGTATCGAACCCCTCCAATGACGCAACCTCCAAACGTGCAAAGGCGGGTTGACGCAAACGTGCAAAATCGAACCACGTGCAAAACGTGCAAAATCGAACCTGACGCAAATCGAACCGGCGGGTGGCGGGTACGTGCAAAACGTGCAAACCTCCAACCTCCAAGGCGGGTTGACGCAAGGCGGGTCCTCCAAACGTGCAAATGACGCAATGACGCAAATCGAACCACGTGCAAAATCGAACCGGCGGGTACGTGCAAACCTCCAACCTCCAATGACGCAAACGTGCAAACCTCCAAGGCGGGTGGCGGGTTGACGCAATGACGCAAACGTGCAAAATCGAACCTGACGCAAGGCGGGTGGCGGGTCCTCCAACCTCCAACCTCCAATGACGCAACCTCCAAACGTGCAAAACGTGCAAAGGCGGGTATCGAACCTGACGCAACCTCCAACCTCCAAACGTGCAAAGGCGGGTCCTCCAAGGCGGGTTGACGCAATGACGCAATGACGCAAATCGAACCTGACGCAAATCGAACCTGACGCAAGGCGGGTATCGAACCCCTCCAACCTCCAAGGCGGGTTGACGCAAACGTGCAAAGGCGGGTATCGAACCTGACGCAACCTCCAAATCGAACCTGACGCAAGGCGGGTTGACGCAAACGTGCAAAGGCGGGTTGACGCAAATCGAACCGGCGGGTATCGAACCGGCGGGTATCGAACCGGCGGGTACGTGCAAAATCGAACCATCGAACCGGCGGGTACGTGCAAA',\n",
    "     11\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is how we can try to find Pattern in relatively small Texts and Strings. For example, when we search for possible interesting patterns in origin of replication.  ''bacterial DnaA boxes are usually nine nucleotides long''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ReverseComplement(String):\n",
    "    '''\n",
    "    Function to return reverse complement of given DNA strand at 5->3 direction\n",
    "    '''\n",
    "\n",
    "    nucleotidesDict = {\n",
    "        \"A\": \"T\", \"T\": \"A\", \"G\": \"C\", \"C\": \"G\"\n",
    "    }\n",
    "\n",
    "    reverseComplement = str()\n",
    "    for i in String:\n",
    "        reverseComplement += nucleotidesDict[i]\n",
    "    \n",
    "    return reverseComplement[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TGATCAAGA'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ReverseComplement(\n",
    "    'TCTTGATCA'\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For example we found some candidates with 9 nucleotides length in baterial Ori. Now we can check, if there are any reverse complement among them."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we need to find pattern coordinates in string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PatternMatching(Pattern, String):\n",
    "    '''\n",
    "    Function to output pattern positions in string\n",
    "    '''\n",
    "\n",
    "    totalCoordinate = 0\n",
    "    positions = list()\n",
    "\n",
    "    for i in range(PatternCount(String, Pattern)):\n",
    "        '''\n",
    "        We need to find match, then move to the matching index plus one to exclude repeating match, but to allow overlapping match\n",
    "        '''\n",
    "        \n",
    "        startCoordinate = String.index(Pattern)\n",
    "        totalCoordinate += startCoordinate\n",
    "        positions.append(totalCoordinate)\n",
    "\n",
    "        String = String[startCoordinate+1:]\n",
    "        totalCoordinate += 1\n",
    "\n",
    "    print(*positions)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17 47 54 109 116 123 147 161 168 198 216 224 267 274 292 299 318 448 455 490 513 589 618 625 632 693 700 756 787 796 803 921 965 983 1010 1017 1043 1087 1111 1148 1262 1281 1304 1392 1399 1406 1426 1444 1453 1460 1476 1491 1606 1613 1706 1713 1740 1764 1810 1817 1824 1847 1895 1916 1934 2082 2140 2176 2202 2309 2378 2400 2417 2466 2507 2514 2571 2578 2606 2615 2649 2670 2753 2760 2790 2828 2858 2882 2897 2923 2954 2970 2977 3011 3018 3033 3082 3089 3136 3233 3240 3261 3279 3286 3293 3431 3511 3606 3629 3645 3723 3810 3825 3832 3848 3875 3882 3890 3897 4001 4119 4144 4151 4181 4214 4230 4315 4390 4397 4433 4463 4470 4481 4496 4511 4540 4550 4571 4587 4639 4646 4653 4678 4693 4735 4742 4846 4853 4863 4943 5005 5012 5028 5035 5052 5059 5083 5146 5154 5169 5176 5188 5215 5248 5287 5297 5326 5361 5368 5375 5383 5398 5424 5449 5456 5479 5486 5503 5554 5571 5655 5695 5722 5739 5789 5796 5803 5821 5846 5868 5876 5883 5923 5942 5960 5967 5982 5989 6059 6097 6104 6208 6224 6231 6238 6245 6282 6333 6379 6404 6432 6447 6454 6479 6515 6523 6531 6557 6565 6572 6579 6586 6609 6625 6632 6639 6657 6664 6778 6802 6812 6925 6932 6939 6973 6981 6988 6996 7003 7041 7071 7178 7185 7203 7218 7225 7232 7278 7363 7392 7401 7439 7467 7474 7601 7626 7643 7674 7705 7805 7824 7831 7838 7872 7926 7941 7978 8004 8012 8032 8049 8057 8088 8174 8216 8314 8341 8375 8434 8441 8490\n"
     ]
    }
   ],
   "source": [
    "PatternMatching(\n",
    "    Pattern='CCTGCCACC',\n",
    "    String='AATCCTGCCAAGAGGGTCCTGCCACCTGCCAATCACCTGCCACTAGGCCTGCCACCTGCCACCCCTGCCAACCTGCCATTCACCTGCCAAAAACCTGCCAACCTGCCAGCCTGCCACCTGCCACCTGCCACCTGCCAGAGCTATCATCCTGCCACCACGTCCCTGCCACCTGCCACCTGCCACGGCCTGCCAAAGTGACCTGCCACCTGCCACTTACCTGCCACCCTGCCACCTGCCAGTCCTGCCATTCCTGCCAAAGCGCCGTGACCTGCCACCTGCCACCTGCCATAGGCCTGCCACCTGCCACCGACCTGCCAACCTGCCACCTGCCAGCGTGATAGCCTGCCAGCCTGCCACATCCTGCCAGCTGTCCACCTGCCAACCTGCCAATATCCTGCCAACGCCTGCCAAGAACAAAACCTGCCATCCTGCCATTAAGCCTGCCAGCCCTGCCACCTGCCACCTGCCATATTGATCCTGCCACAGGCCTCCTGCCACCTGCCAGCTATATTGCCTGCCACCTGCCAAACAGGGAGACCTGCCACGACCTGCCATCCTGCCAGCCTGCCACTTTAACCTGCCATGGGGCCCTGCCACCTGCCAACGCCTGCCACTCGGCCTGCCACCTGCCACCTGCCACCTGCCAGACAGCCTATCCTGCCAACCTGCCATCCTGCCATCCTGCCAAGAAAGCCTGCCACCTGCCACCGCCCGTTGTCCTGCCACGCCCTGCCAGCCTGCCAAAGCCTGCCAAGACCTGCCACCTGCCAGCCTGCCATTTTCTTCTCCTGCCACCCCTGCCACCTGCCACCCTGCCATTCCTGCCATGGCGCGAACCTGCCAACCCTGCCACTACCTGCCAGCCTCACTCCTGCCAGCCCTGCCAGTCGCCTGCCAAGTAACGCCTGCCAATTACCTGAGCCTGCCACCTGCCATCCTCCTGCCAGGCCTGCCACGCCTGCCAACCTGCCACCTGCCAGGGCCCTGCCACCTGCCAAACCTGCCAACGCCCTGCCACCTGCCACCGCCTGCCAGGCTGCGACCCTGCCACCCTGCCATCCTGCCAGCCTGCCAGTCCTGCCAAGTTCCTGCCACCTGCCAGCCTGCCACTCCTGCCACCTGCCACAAGGTGGTCCCTGCCAGTTATACCTGCCACCTGCCAAATTCCCTGCCACACCCTCCCTGCCATCCTGCCACTAGTGCCAGTATGCCTGCCAATCCTGCCAGTGCCCTGCCAACCTGCCATCGGTGACCCTGCCAAGCCTGCCACCTGCCACTTTTCCTGCCACCCTGCCAACGATCGACCTGCCACCTGCCACTTAGGTCCCTGCCAGAGCCCTGCCATGGACCTGCCACAGCCGGTCCCCCTGCCATCCTGCCACGCCTGCCACGCCTGCCACCTGCCACCTGCCACCTGCCATCCGCACCTGCCACCCCTGCCAGCCCTGCCACCCCTGCCACCTGCCACCTGCCATTCCTGCCACCTGCCATCCTGCCACCTGCCATCCTGCCATACCTGCCAACCTGCCAATCCCTGCCACTTGCCTGCCAATAACGATGCCACCTGCCACATTCCTGCCAACCTGCCAGCTTCCCTGCCATGAGACCTGCCACCTGCCACCTGCCATCTTCCTGCCATCCTGCCAACCTGCCAACGATAACCCCTGCCAGGCTCCTGCCAGAGCCTGCCAATCCGATCCTGCCATCCTGCCACCTGCCACCTGCCATCGTCCTGCCATACCTGCCACCTGCCATCCTGCCAGTCCTGCCACCAATGGGACCTGCCATTCCCCCTGCCACTCGAGTTAGCCCTGCCACCTGCCACCTGCCACCTGCCAGCCTGCCAGCCTGCCACCTGCCAGCGAAATACCCTGCCAAGCCCTGCCATTGGCACACCTGCCACCTGCCAACCTTCACCTGCCACCCTGCCATGTCCTGCCACCGCGAATCCTGCCAAAGTGGACCTGCCAGCGGGGTAGTCCTGCCAAACCCTGCCAGTAATCACACCTGCCATCCTGCCATAATCCTGCGAGCCTGCCATGCAAGCCTGCCAGCCTGCCACTCCTGCCAACCCTGCCACAGCCTGCCACCTGCCATCCTGCCAATGAACCCTGCCATGTTCGCCCCGTAGACCTGCCATCCTGCCACCTGCCAGCCTGCCAAGAATCCTGCCAGCCCTGCCACCTGCCATGTCCTGCCATACCTGCCACCTGCCATAACCTGCCATCCCTGCCACGCACCCTGCCAGCCTGCCAATGCGCACCTGCCAACCTGCCAGGTATGCCTGCCAGAGACCAAGACCTGCCAACCCTGCCACCTGCCAACCTGCCATGATCCCTGCCATCCTGCCAATGTCCCTGCCATTCCCCTGCCAGGAACCTGCCACCGTCCCTGCCAAGTCCTGCCACCTGCCAAATCCTGCCACCTGCCATTACCTGCCATAACCTGCCAATTCCTGCCATGACTCCTGCCACCACCTGCCAGCACCTGCCAGACCTGCCACTGTCCCTGCCACCTGCCACCTACCTGCCAGCTGAAACCCCACCTGCCCTGCCAGCTACACCTGCCATGCCTGCCACCTGCCACCTGCCAACCCTGCCAGTAACCCTGCCACCCCTGCCACCCTGCCATTCCTGCCAGATCAACTCACCTGCCACCATACCTGCCAAGCCTGCCACCTGCCAACGCCTGCCAGGGCGTCCTGCCACGGCCTGCCATTCCTGCCAGGCGTCCTGCCAGTGGGCCTGCCAGTTCCTGCCACCTGCCACCTGCCAACAACCTGCCAGGAGGCCTGCCACCTGCCATGACCCTGCCATCCCTGCCACATGCCTGCCACCTGCCAGCAGTCCTGCCATACTCCTGCCACCTGCCAATCCTGCCAGCCTGCCACCTGCCAGCCTGCCACCTGCCAGCCTGCCAATGACCTGCCACCCCAGCACTGCCTGCCACAGGTCCCTGCCACCCTGCCAGCCTGCCACCTGCCACCCTGCCAGACCCTGCCAAGTACGCCCCCTGCCACCTGCCACCTGCCAGCCTGCCACCTGCCAACTCAAGCCTGCCAGATAGTACAGACCTGCCAGATCCTGCCACCTGCCACCTGCCACAGAAGAGAAAACCACCTGCCACACCTGCCATTCCTGCCACCTGCCAGCGCCTGCCAAGTTCTGTACCTGCCAGCCCTGCCAATCCTGCCAAGCCTGCCATCCTGCCACGCGTCTCAGACCTGCCAACTCCCTGCCACCTGCCACCTGCCATTGCCCGCCTGCCACCTGCCAATACCCTGCCACCTGCCACCTGCCACCACCTGCCAAGAGTCCTCTCCTGCCACTGGCCTGCCAGCCTGCCATACCTACCTTTACCTGCCATCCCTGCCAGTGCAATCCTGCCAACCTGCCAAACCCCAATGGTTCCTGCCATCGGAACCTGCCACACCTGCCACCACTTCCTCCCTGCCACAGCTTTACCTGCCATTCCTGCCATCCTGCCAATCGCCTGCCATACCTGCCAATGACCTGCCACCTGCCAACCCCTGCCAATTCCTGCCATATCACCTGCCATGGTTCATCCTGCCACTCCCTGCCAGAGCCAACCTGCCAACCTGCCATCCCTGCCACCTGCCAAGCGGAGTTCCTGCCACCTGCCAGTCCTGCCACCTGCCATGGAGTGCCTGCCAATGCCTGCCACGTCCTGCCAAGCCCTGCCATTTCCTGCCAGCCTGCCATGCCTGCCACCCTGCCAATTAGCGCCTGCCATCGATCCCTGCCACTGGCGTTCCTGCCAGGGTCCTGCCAACACACCCCCCTGCCATGACCTGCCACCTGCCAACCTGCCACCTGCCACCTGCCAGTCCTGCCACCCTGCCATGTATCACTGAGCCTGCCACCTGCCACCCTGCCACCTGCCACCTGCCAACCCCTGCCAATCCTGCCAGACCTGCCAGTAAGCCTGCCACTCCTGCCAACTAACCTGCCACGATTCCCTGCCACTGACTCCTGCCAGCCCCTGCCACCTGCCAAACAAACCTGCCAACCTGCCAGGACCTGCCAAACCTGCCATAACGACCCCGTGTGATCCTGCCAGCCTGCCATCCTGCCAAGGCCTGCCAACCTGCCAATGTCACCTGCCACCTGCCAAACCTGCCAATCCTGCCACCTGCCACCTGCCAGGCCGGCGCCTGCCAACCTGCCACCTGCCAACCTGCCAGCCCTGCCATACCTGCCACCTGCCAGCCCTGCCACCTGCCAGCCTGCCAACCTGAACAGCCTGCCATCCTGCCAACCTGCCAGACCCCTGCCACTTAAATATTCCCTGCCATCCTGCCACCTGCCAACCTGCCACGTACCTGCCAGCTGCGCCCGGGCCTGCCATCAAAGACCCTATGTCCTGCCATCCTGCCACCTGCCACCCCTGCCAACCTGCCATGTCCTGCCAGCCCTGCCACCTGCCATGGCCTCCTGCCAAGGCCTGCCACCTGCCACCTCCCTGCCACCTGCCAACCTGCCACCCTTGTTCCTGCCACCTGCCAGCACCTGCCAGATGGCCTGCCACCCCCTGCCACCTGCCATTGTCTACCTGCCACCTGCCAGTCCTGCCACCTGCCAGTCCTGCCAGGCGCCCTGCCATCCCTGCCACGTTCACACCTGCCACCTGCCACCTGCCACCTGCCATCCTGCCACGGCCTGCCACCTGCCAGCCTGCCACCTGCCATCTAATAACCTGCCATCCCTGCCAATAACCTGCCACCTGCCACCTGCCAGTTACCTGCCAAATATATCATCGAACGTGAGTCTTCCGGACAACGCCTGCCACTATCGTCGCCTGCCAGCCTGCCATCCCCTGCCAAATGCCTGCCACCTGCCACCACCTGCCACCTGCCACGAGGGTAAATCCTGCCAATTTAGATCGGCGTCCTGCCATGCTTGGACCTGCCATACCCTGCCAAGCCTGCCACCTGCCATTGCCTGCCAATGCTAGCCTGCCATTGCCTGCCAAAGAACAGTTCCCACCTGCCACCTGCCACCTGCCACTCCTGCCACCTGCCACCTGCCAAATCCTGCCACCTGCCACCTGCCATTGGCCGCATCCTGCCACCTGCCAGATCCTGCCACTTCCTGCCAGCGTCCCTGCCACACCTGCCAATAGCACACCTGCCACCCTGCCACCTGCCAACCTGCCACCTGCCACCAATCCTGCCACCTGCCAGCACCTGCCACAACCTGCCACCTGCCAAATAACCTGCCATGTGTTACCTGCCACCCTGCCAGAACAGAAGACGCACCCTGCCAATCCTGCCACCCCCTGCCACCAGCCTGCCAACCCTGCCACACCTGCCACCTGCCAGCCTGCCATCCTGCCATTTTACCTGCCACCTGCCACCTGCCACCCTGCCACCTGCCATCCTGCCACCCATGTCCTGCCAGATGGCCTGCCACCTGCCAGGTCACAACCTCCTGCCACCTGCCACCTGCCTCATGAAAAGCCTGCCACCTGCCACCTGCCAGATCCTGCCACCCTGCCATCCTGCCAACCCAAGGCCTGCCATACACCTGCCAAACCTGCCACCTGCCACTTCCTGCCACCTGCCATTTCACCCTGCCATTTCTGGCGGAGTCGTGTCCTGCCAGGGCGTCCTGCCACGGACCTGCCAGCGCGGTCCCTGCCACCTGCCAAGTTGGCCTGCCAACCTGCCAAGAGCCCTGCCACCCCCTGCCAACCTGCCACGCCTGCCACCCTGCCAACCCTGCCACCCTGCCACAGGCCTGCCAGCTTACCTGCCATCCTGCCAGCAGCCTGCCACCTGCCACCTGCCACCTGCCATCGGCCTGCCACCTGCCAACCTGCCATACCCTGCCACCCTGCCACGGTCCTCCTGCCACCCTGCCACCTGCCACCCTGCCAGCAGGAAGATCCTGCCAGAACCGCCCCTGCCACCACCTGCCACACCTGCCACCCTGCCACTCCCTGCCACCTGCCACCTGCCAGCCTGCCACCTGCCACCTGCCATCCTGCCATCCTGCCAACTCCTGCCATGCCTGCCACTCCTGCCATACTCCTGCCATCCTGCCACCTGCCAACCTGCCAGTCCCTGCCAGGGCGACCTGCCACCTGCCACCTGCCAGGTGCCTGCCACACCTGCCACTCCTGCCAGTGGTACGCTTTTCCCTGCCATAATCCTGCCATCCTGCCACGTTGCCTGCCATCCTGCCAACCTGCCACCTGCCACTCCTGCCACCTGCCACCTGCCACCTGCCACCCTGCCACGTCCAACTGTTGCCTGCCAGACCTGCCACCTGCCAAGTAGCCCTGCCAGAGCCTGCCAGGCCTGCCATTTGTCCTGCCACCTGCCAGTTAGCTCCTGCCATGCAATCCCTGCCATGAACCTGCCACCTGCCAACCTGCCACGCCCTGCCACCCTGCCACACTCCTGCCATCCCTGCCACCTGCCATCCTGCCACCTGCCACCCTGCCAGCCTGCCAAACCTGCCACCTGCCATGCACCTGCCATTTCCTGCCATCCTGCCACCCTGCCACCCTGCCACCTGCCAACCCTGCCAGGCCCTGCCACCCTGCCACCTGCCACCTGCCACCTGCCACCCAAGCCTGCCACTTCCTGCCACCTGCCACGCCTGCCACCTGCCACCTGCCACCTGCCAGCTGCCTGCCACCTGCCACCTACCTGCCATGGACTAGCGCCTGCCAGTCCTGCCAAGAATCTCCCTGCCATTCAAGGTCCTGCCAATCACGGGTAGCCTGCCAAAGCGTCATTTCCTGCCAATTCCCTGCCACCTGCCAAACCTGCCATCCTGCCACCTCCTGCCACCCTGCCAGCAATCCTGCCAATGCCTGCCAACCCTCCTGCCAGCCTGCCAGAAACCTGCCAGCTCCTGCCACATCCTGCCAATCCTGCCAGAAAACCCTGCCAAGCCCTGCCACCTGCCACCTGCCACCTGCCAACCGATGACTTCCTGCCATTCCTGCCACCCTGCCACCTGCCACCCTGCCACCTGCCACCTGCCATACCTGCCAAACAAAACCTGCCAGCCTGCCACCCTGCCATCGCCGCCTGCCATGCCTGCCACCTGCCACAACCTGCCAATTCCTGCCACATTCCTGCCATCCTGCCATCCCTGCCAGACTACCCACCTGCCACAAAAACCTGCCATTACCCTGCCAACCAACCTGCCACCTGCCACCAGCAAATCACCTGCCACCTGCCAGCCTGCCACCTGCCACCTGCCACCTGCCAGGCCTGCCAATCCTGCCATCCCTGCCATACGGCCTGCCACCTGCCAGGGATCCTGCCAGAACTTTGGTTCGGGCCTGCCATCCCTGCCATTCACGTCCTGCCAACCTGCCACTCGCACCTGCCACCTGCCACGCCTGCCACTAGACCCTGCCACCCCTGCCACCTGCCACAACCTGCCATGACCTGCCAGTCTCCTGCCACCTGCCAATCCTGCCAACTCCCCTGCCACCTGCCACCACCTGCCATAGTCCTGCCAGGACCTGCCACTGGTACCTGCCATCCTGCCAAGAGCCAACCCTGCCATGCCCTGCCATGACCTGCCAGCCCCTGCCAATGTCCTGCCACTGCCTGCCATCCTGCCACCCTGCCATCCCTGCCATCCTGCCACCTGCCATCACCTGCCACCCTGCCACAAGCCTGCCACTTTCCCTGCCACCTGCCAACCTGCCATCCTGCCAACCTGCCACCTGCCATATACGAACCTGCCACAGCAAGCCCTGCCAGCCCTGCCATACCTGCCACTTCCTGCCACAGTGTCATCTGTCCTGCCAACATGTCACCTGCCACCTGCCAAAAACCCTGCCACCTGCCACCTGCCACCTGCCATCCTGCCATTGAACTCAATCCCTGCCACCTGCCATCCTGCCACTCCTACATCCTGCCAGCCCTGCCATTTGGCACCTGCCACCTGCCAACCTGCCACCCTGCCAGCCTGCCAGACCCTCCTGCCATCCTGCCACCCTGCCACGCACCGCGCACCTGCCACCCTGCCACCCTGCCACATGTCCTGCCACCTGCCATGCCCTGCCACCCTGCCACCTGCCATGAACCTGCCACGCGCGCCTGCCACCTGCCAGCTCCTGCCAGCCTGCCAACGCCCTGCCACTTGCCACCAAACCTGCCATACTTCCTGCCAAAAACCTGCCAACCTGCCACCTGCCAGACCCCCTGCCATAAACTCTCCTGCCAACCTGCCACCTATAGATCCTGCCAACCTGCCATGACTCCTGCCAACCTGCCAAGTCTCGCAAACCTGCCACAAGGGGCCTGCCAGCTGACCTGCCAATGCCTGCCACCTGCCAAGAGTTCCATCAACCTGCCACCTGCCAGCCTGCCACTGTCCTGCCAGCCTGCCACCCTGCCACTTTCCTGCCATTGCCTGCCAGATGGGTGCCACCTGCCATGTCACCTGCCACCTGCCACCTGCCAACCTGCCACGTCCTGCCAGGTCGAACCCCTGCCATCCTGCCACCTGCCATGCCTGCCAGCCCTGCCAAACTCCCGGTCCTGCCAACCTGCCACGGCCTGCCACATACCCCCTGCCAACCTGCCAGAACCTGCCAATCCTGCCA'\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Попробуем найти все стартовые позиции нашей 9-ти нуклеотидной последовательности CTTGATCAT в геноме V.cholerae"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Users/olegsuchalko/BioinformaticsAlgorithms/Part_1/Vibrio_cholerae.txt') as vibrio_cholerae_genome:\n",
    "    genome = str(vibrio_cholerae_genome.readline().strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "60039 98409 129189 152283 152354 152411 163207 197028 200160 357976 376771 392723 532935 600085 622755 1065555\n"
     ]
    }
   ],
   "source": [
    "PatternMatching(\n",
    "    Pattern='CTTGATCAT',\n",
    "    String=genome\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "116556 149355 151913 152013 152394 186189 194276 200076 224527 307692 479770 610980 653338 679985 768828 878903 985368\n"
     ]
    }
   ],
   "source": [
    "PatternMatching(\n",
    "    Pattern=ReverseComplement('CTTGATCAT'),\n",
    "    String=genome\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Для паттерна и его комплементарной последовательности находим участок, где оба варианта встречаются с повышенной плотностью, нежели чем в остальном геноме. Значит мы хотим искать участки генома, где эти паттерны часто встречаются (clumps) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def FindClumps(Text, k, L, t):\n",
    "    '''\n",
    "    Function to find clumps of k-mers in window of size L more than t times\n",
    "    '''\n",
    "\n",
    "    Patterns = set()\n",
    "    n = len(Text)\n",
    "\n",
    "    for i in range(n - L):\n",
    "        Window = Text[i:L+i]\n",
    "        # print(Window)\n",
    "        freqMap = FrequencyTable(Text=Window, k=k)\n",
    "        # print(SortPatternsDict(freqMap))\n",
    "        \n",
    "        for key in freqMap:\n",
    "            if freqMap[key] >= t:\n",
    "                Patterns.add(key)\n",
    "    \n",
    "    return Patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CGACA', 'GAAGA'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FindClumps(\n",
    "    Text='CGGACTCGACAGATGTGAAGAACGACAATGTGAAGACTCGACACGACAGAGTGAAGAGAAGAGGAAACATTGTAA',\n",
    "    k=5, L=50, t=4\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TCCAGTCCAG'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FindClumps(\n",
    "    Text='CAGATATCAACTCTTAGCCACTTGACTTATGCTGTTGTATCTAGCGCACCAGTTTCTAGCCAAGTGTTAGTACCGGGGAGTATCGGTATGAAATAAGCGTTTACGGAAAGAAGAGGCCCAGCTCCCAGCTAGCTCTAGAGGAGGCGCTCTAGAGGGAGGTATGAGTGGGAGATGAACCCACGGCTCCCGTATCCGAATAGTTTCTTACCTTAACCTCTTGACCTAAAGTGGATGAGAGCCTCCTGGGCGATCCATTTAACCGCTTCACTGATAACTTCCTCCCGAAGTTAGGCACTGGGCGCGGGGTATTAGTACAGTGTTGTAATCACTGTTGAACGCTCCACTCAATGCGTTACGCAGGCCGCCGACATCTCGTGATGGTTCGTTCGGTTCTTCACTCATTAGAGTATGGCCCCCTCAAATGCACTGAGGTCACGAGAATCGTTGATCTGATTGCGAGGACAAGGGTTTGTTGAACAGTCTAGCAAGCACTGCAAGTGGAAACCAAATGACTCCAGTCTGGCACCTGTTACGTGCCTGCCACACGGAGCCGATCCGCGCGCGTAGTGCTAGTGCGCGTAGTGCGCCCAAAAGGGGGTCGGACCGGGGGATTCCCCCGCTATAATACGTCTCATGGAGGTGGCGGGTACCGAGGGGCAGCACAAATAGAGGTCTCTCGATCCCACCGTATGGGAAAGAAAATATACTGCGGGTATCGCGCCAGGTTATAGAAGCGCCTTAGGGCAGGTAGTCCCGTAAACGACTTCGGATAGATCTTATCGGGCTTGTGGGCGATCCTACGTCTAGCGATTATTGCCCCCAGTCAACCGCAATCCTAGCGGGTGATAAGTATCATGGCTTAGTACAACGAAGCCTGAATTAAGGCTCAATCTTTACATAATCAATCTTTACATCCCGCCCCGCCCGCCGCCCGTCTTCCCGCCCGTCTGACTTAAGGTTTTAAGAGGTAAGTGATCAGATACTTCCGGCACTGACTTATCCAACGTCATCATCCAACGTCCTCCTGCGCTAACTTTGCGGATTTTTAGAGCGACGCAAGGTATCCTCGCGAGCATCCGGAAGCACTCGCATACTCGGGGTATTAACATGTCTGGCGTATGAATAGACCAATTTGCAGACTAGCAAACCAACCTGTCTGAACTACGGGTGCCGAAAGGCTTAACCTTACCTTGAGTCCGGGCAATGGCGGCAGCTAGTGCTCCCTACTTTGACTTTGACTCGTAACTATCGTTCACTTCACTCCAAAATCACTCCAAAAAACTGGGTAGTAATAATTGGAAGCATCAGGCGGAGAACCGTCGGACTCACCCACCGTTCCAGTCCAGTCCAGTCCAGAACAGTCCAGAAGTCAAAAGTCAAAAGTCAAAATCCTTACAAAACTCGCCCTGTGCCTAAGAACGTGCTAGAAGAGGGAAGTATTGGGCAGAGTAGACTACGCGAGCACGGGAAGAGTATCACTATCACTGCTAACTGCTAGCTAGGAGACCGATGGTGGGCCCGCAGCATCTAGATAAAACGAAAGTCGGACTTCCATCAATCGTTGCCCTAAAAGCGTGGGAACTCTTGAGGCGCTGTCGTTCGCATCTCGTCCGTATGTATGCGCCGTATGCGCCCCGCCCGTCATGGGGTATAGGATCTGGGCATCTATTTTAAGTGCGGCACGCTCCCACCCAGCGTGTGATCAGATAGACTGTTCGGGCGGTAGAGCGGCCGACTCGTTGATCTCCGGGAGCTCTTTCATGCAATGCTAAAAATTGTCTGGTTTCATTGGTGGGACTCTACGGTAAGGCTTAATTAAGGCTTAATAAAACGGTCAGTCTCTAGCGAATTAGATGCAACAAACAGATATTGCAGTTGAGTCCGCCTGGAGGTGCGTGGAGCAACAACCACACCCCAACCACACCTGACAATGTTGGATATCATCTCCCGCACAACAACTACATGGTGATCTGTCAATTAAACTAAATCAG',\n",
    "    k=10, L=24, t=3\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"/Users/olegsuchalko/BioinformaticsAlgorithms/Part_1/E_coli.txt\") as genome_file:\n",
    "    ecoli_genome = genome_file.readline()\n",
    "\n",
    "# FindClumps(\n",
    "#     Text=ecoli_genome,\n",
    "#     k=9, L=500, t=3\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = SortPatternsDict(FrequencyTable(Text='TAAACGTGAGAGAAACGTGCTGATTACACTTGTTCGTGTGGTAT', k=3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AGCTAGC'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ReverseComplement('GCTAGCT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 1 12\n"
     ]
    }
   ],
   "source": [
    "PatternMatching(\n",
    "    String='AAACATAGGATCAAC', Pattern='AA'\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12 (main, Apr  5 2022, 01:53:17) \n[Clang 12.0.0 ]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "40d3a090f54c6569ab1632332b64b2c03c39dcf918b08424e98f38b5ae0af88f"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
